overlord/ifacestate/ifacemgr.go

```diff
@@ -304,26 +304,62 @@ func (m *InterfaceManager) ConnectionStates() (connStateByRef map[string]Connect
 // In both cases the snap name can be omitted to implicitly refer to the core
 // snap. If there's no core snap it is simply assumed to be called "core" to
 // provide consistent error messages.
-func (m *InterfaceManager) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slotName string) ([]*interfaces.ConnRef, error) {
-	connected := func(plugSn, plug, slotSn, slot string) (bool, error) {
-		_, err := m.repo.Connection(&interfaces.ConnRef{
-			PlugRef: interfaces.PlugRef{Snap: plugSn, Name: plug},
-			SlotRef: interfaces.SlotRef{Snap: slotSn, Name: slot},
-		})
-		if _, notConnected := err.(*interfaces.NotConnectedError); notConnected {
-			return false, nil
-		}
+func (m *InterfaceManager) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slotName string, forget bool) ([]*interfaces.ConnRef, error) {
+	var connected func(plugSn, plug, slotSn, slot string) (bool, error)
+	var connectedPlugOrSlot func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error)
+
+	if forget {
+		conns, err := getConns(m.state)
 		if err != nil {
-			return false, err
+			return nil, err
+		}
+		connected = func(plugSn, plug, slotSn, slot string) (bool, error) {
+			cref := interfaces.ConnRef{
+				PlugRef: interfaces.PlugRef{Snap: plugSn, Name: plug},
+				SlotRef: interfaces.SlotRef{Snap: slotSn, Name: slot},
+			}
+			_, ok := conns[cref.ID()]
+			return ok, nil
 		}
-		return true, nil
-	}
 
-	connectedPlugOrSlot := func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error) {
-		return m.repo.Connected(snapName, plugOrSlotName)
+		connectedPlugOrSlot = func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error) {
+			var refs []*interfaces.ConnRef
+			for connID := range conns {
+				cref, err := interfaces.ParseConnRef(connID)
+				if err != nil {
+					return nil, err
+				}
+				if cref.PlugRef.Snap == snapName && cref.PlugRef.Name == plugOrSlotName {
+					refs = append(refs, cref)
+				}
+				if cref.SlotRef.Snap == snapName && cref.SlotRef.Name == plugOrSlotName {
+					refs = append(refs, cref)
```

Comment:
this path doesn't seem to be covered by unit tests

Reply:
Test added.